### ISSUE TITLE
plugin/cache: prefetch doesn't actually prefetch

### ISSUE THREAD

**Issue by grobie** [coredns/coredns#1278]
We're about to consolidate our DNS infrastructure and plan to use CoreDNS everywhere (authoritative servers, resolvers, caches, forwarders). SoundCloud relies on DNS heavily for service discovery and quick DNS lookups are key for the overall latency.

While doing some load testing, I noticed that the prefetch option requested in #581 doesn't prefetch the record in the background, but blocks actual queries. This differs from implementations in other popular DNS servers (like bind9) and doesn't help to ensure fast lookups in case authoritative/upstream servers are slow. With the CoreDNS implementation, the prefetch option will simply cause an earlier request to become bound by the upstream latency, instead of allowing all queries to be answered from the cache.

> The intention is that the fresh copy would arrive before the existing copy expires, which ensures a uniform response time.

https://deepthought.isc.org/article/AA-01122/0/Early-refresh-of-cache-records-cache-prefetch-in-BIND-9.10.html

Question: Is this intentional? I'm happy to contribute a change to prefetch a soon-to-expire cache entry in the background (go routine) and ensuring uniform response times.

**Comment by johnbelamaric** [comment]
Um. No, I am pretty sure that was not intentional...

**Comment by grobie** [comment]
I started working on a new prefetch implementation.

**Comment by miekg** [comment]
Just returning (the maybe somewhat stale) entry from the cache, while looking up the new value might be a relative small change?

Also in handler.go, if the response was in the cache it is written to the client immidiately, after that happened the prefetch stuff is checked. So why would this block new queries?

**Comment by grobie** [comment]
> So why would this block new queries?

It was a combination of multiple issues leading to my false statement, mostly that prefetching didn't work at all in the default config for answers with low TTLs (<10s), and that both logs and metrics reported wrong response times for requests triggering the prefetch.

**Comment by grobie** [comment]
I finally had some time to pick this story up again. I fixed a couple of issues with prefetching, see #1363. It's far from being perfect, but it provides the minimum changes we need to use CoreDNS as caching server.

In general, the cache implementation would benefit greatly from some refactoring and decoupling of concerns. I'll see if I can find some time to clean up the current state.

I didn't touch the general prefetch approach, but it prevents a lot of our requests from being prefetched. We have clients which cache again on their end. These clients request the record all in ~1-2s at the begin of a new TTL interval, and then only retry once the TTL has passed. This means there are no requests which can trigger a prefetch. This could be solved by scheduling goroutines to only run shortly before a TTL expires. As the latency of requests made by such caching clients is not important, I refrained from implementing a more advanced prefetch approach.
